util/tracecpu/collector.go

```diff
@@ -0,0 +1,333 @@
+		delete(ts.topSQLMap, evict.Key)
+		sqlDigest, planDigest := decodeCacheKey(evict.Key)
+		ts.mu.Lock()
+		delete(ts.normalizedSQLMap, sqlDigest)
+		delete(ts.normalizedPlanMap, planDigest)
+		ts.mu.Unlock()
+	}
+}
+
+// RegisterNormalizedSQL registers a normalized sql string to a sql digest, while the former can be of >1M long.
+// The in-memory space for registered normalized sql are limited by TopSQL.normalizedSQLCapacity.
+//
+// This function should be thread-safe, which means parallelly calling it in several goroutines should be fine.
+// It should also return immediately, and do any CPU-intensive job asynchronously.
+// TODO: benchmark test concurrent performance
+func (ts *TopSQLCollector) RegisterNormalizedSQL(sqlDigest string, normalizedSQL string) {
+	ts.mu.RLock()
+	_, exist := ts.normalizedSQLMap[sqlDigest]
+	ts.mu.RUnlock()
+	if !exist {
+		ts.mu.Lock()
+		ts.normalizedSQLMap[sqlDigest] = normalizedSQL
+		ts.mu.Unlock()
+	}
+}
+
+// RegisterNormalizedPlan is like RegisterNormalizedSQL, but for normalized plan strings.
+// TODO: benchmark test concurrent performance
+func (ts *TopSQLCollector) RegisterNormalizedPlan(planDigest string, normalizedPlan string) {
+	ts.planRegisterChan <- &planRegisterJob{
+		planDigest:     planDigest,
+		normalizedPlan: normalizedPlan,
+	}
+}
+
+// this should be the only place where the normalizedPlanMap is set
+func (ts *TopSQLCollector) registerNormalizedPlanWorker(planDecoder planBinaryDecodeFunc) {
+	// NOTE: if use multiple worker goroutine, we should make sure that access to the digest map is thread-safe
+	for {
+		job := <-ts.planRegisterChan
+		ts.mu.RLock()
+		_, exist := ts.normalizedPlanMap[job.planDigest]
+		ts.mu.RUnlock()
+		if exist {
+			continue
+		}
+		planDecoded, err := planDecoder(job.normalizedPlan)
+		if err != nil {
+			fmt.Printf("decode plan failed: %v\n", err)
+			continue
+		}
+		ts.normalizedPlanMap[job.planDigest] = planDecoded
+	}
+}
+
+// snapshot will collect the current snapshot of data for transmission
+// This could run in parallel with `Collect()`, so we should guard it by a mutex.
+func (ts *TopSQLCollector) snapshot() []*tipb.CPUTimeRequestTiDB {
+	ts.mu.RLock()
+	defer ts.mu.RUnlock()
+	total := len(ts.topSQLMap)
+	batch := make([]*tipb.CPUTimeRequestTiDB, total)
+	i := 0
+	for key, value := range ts.topSQLMap {
+		sqlDigest, planDigest := decodeCacheKey(key)
+		normalizedSQL := ts.normalizedSQLMap[sqlDigest]
+		normalizedPlan := ts.normalizedPlanMap[planDigest]
+		batch[i] = &tipb.CPUTimeRequestTiDB{
+			TimestampList:  value.TimestampList,
+			CpuTimeMsList:  value.CPUTimeMsList,
+			SqlDigest:      sqlDigest,
+			NormalizedSql:  normalizedSQL,
+			PlanDigest:     planDigest,
+			NormalizedPlan: normalizedPlan,
+		}
+		i++
+	}
+	return batch
+}
+
+func (ts *TopSQLCollector) sendBatch(stream tipb.TopSQLAgent_CollectCPUTimeClient, batch []*tipb.CPUTimeRequestTiDB) error {
+	for _, req := range batch {
+		if err := stream.Send(req); err != nil {
```

Comment:
Do we need to assign some batch policy for gRPC, so that this doesn't result in gRPC so send a lot of TCP frames, but to merge multiples into one?

Reply:
I'll look into this.

Reply:
Maybe worth to try with https://pkg.go.dev/google.golang.org/grpc#WithWriteBufferSize

Reply:
duplicated with https://github.com/pingcap/tidb/pull/24988/files/68df7b2aea6d209d1420b6fb8654ae8e7d1e5ea1#r643062144, will handle this there